### PR TITLE
Nullable: System.Collections remainder (non-generic)

### PR DESCRIPTION
**Pull request by krwq**
<s>Do not merge until https://github.com/dotnet/coreclr/pull/24258 is merged (to avoid merge conflicts/issues)</s> (already merged)

### CODE REVIEW

File: src/System.Private.CoreLib/shared/System/Collections/IHashCodeProvider.cs [hunk 1]
```diff
@@ -13,6 +14,6 @@ namespace System.Collections
     public interface IHashCodeProvider
     {
         /// <summary>Returns a hash code for the given object.</summary>
-        int GetHashCode(object obj);
+        int GetHashCode(object? obj);
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Collections/IHashCodeProvider.cs]: This allows null?

File: src/System.Private.CoreLib/shared/System/Collections/Hashtable.cs [hunk 2]
```diff
@@ -290,13 +291,13 @@ public Hashtable(int capacity, float loadFactor)
...
 
-        public Hashtable(int capacity, float loadFactor, IEqualityComparer equalityComparer) : this(capacity, loadFactor)
+        public Hashtable(int capacity, float loadFactor, IEqualityComparer? equalityComparer) : this(capacity, loadFactor)
         {
             _keycomparer = equalityComparer;
         }
 
         [Obsolete("Please use Hashtable(IEqualityComparer) instead.")]
-        public Hashtable(IHashCodeProvider hcp, IComparer comparer)
+        public Hashtable(IHashCodeProvider? hcp, IComparer? comparer)
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Collections/Hashtable.cs]: The ctor below should have ? on the comparer argument, too, right?

**Review by stephentoub — COMMENTED** [review]
A few comments to be addressed, otherwise LGTM. Thanks.

File: src/System.Private.CoreLib/shared/System/Collections/IHashCodeProvider.cs (hunk 1 above)
Comment (krwq) [reply src/System.Private.CoreLib/shared/System/Collections/IHashCodeProvider.cs]: good catch - I think I've seen the implementation which allowed null but now when I looked again there are actually no implementations here (it must have been a very specific implementation of equality comparer) - let's see if we get compile errors when we move it to other repos but for now better to keep arguments more constrained

**Pull request merged**